File: 0400_0499/leetcode-no416.py
class Solution:
    def canPartition(self, nums: list[int]) -> bool:
        s1 = sum(nums)
        if s1 % 2:
            return False
        else:
            s1 //= 2
        nums.sort(reverse=True)
        n = len(nums)
        def dfs(total,i):
            # print(1,"*"*i,total)
            if total == 0:
                return True
            a = False
            # print(2)
            while i < n:
                # print(3,i)
                if total >= nums[i]:
                    a = dfs(total-nums[i],i+1)
                i += 1
                if a:
                    break
            if a:
                return True
            return False
        return dfs(s1,0)

t = Solution()
print(t.canPartition([14,9,8,4,3,2]))
# print(t.canPartition([1,2,3,5]))
# print(t.canPartition([1,5,11,5]))
# print(t.canPartition([4,7,1,6,2,3,9,8,5,10]))
# print(t.canPartition([4,7,1,11,6,2,3,9,8,5,10]))

File: leetcode-cpp/0400_0499/leetcode-no416.cpp
// problem: https://leetcode.cn/problems/partition-equal-subset-sum/description/?envType=daily-question&envId=2025-04-07
#include <iostream>
#include <memory>
#include <numeric>
#include <vector>

#include "../utils/vector_helper.h"

using namespace std;

class Solution {
public:
    bool canPartition(vector<int> &nums) {
        int total = accumulate(nums.begin(), nums.end(), 0);
        if (total % 2) {
            return false;
        }
        auto half = total / 2 + 1;
        vector<bool> bag(half);
        bag[0] = true;
        int n = nums.size();
        for (int j = 0; j < n; ++j) {
            int t = nums[j];
            for (int i = half - t - 1; i >= 0; --i) {
                if (bag[i]) {
                    bag[t + i] = true;
                }
            }
        }
        // show_vector(bag);
        return bag[half - 1];
    }
};


int main(int argc, char *argv[]) {
    Solution test;
    cout << test.canPartition(*make_unique<vector<int> >(vector{1, 2, 5})) << endl;
    cout << test.canPartition(*make_unique<vector<int> >(vector{1, 5, 11, 5})) << endl;
    cout << test.canPartition(*make_unique<vector<int> >(vector{1, 2, 3, 5})) << endl;
}


//
// Created by ASUS on 2025/4/7.
//
